tests/automatic_calibration/test_calibration_controller.py

```diff
@@ -0,0 +1,849 @@
+G7_calls = [call(zeroth),call(zeroth),call(first),call(second),call(zeroth),call(third),call(zeroth),call(first),call(fourth)]
+G8_calls = [call(zeroth),call(second),call(zeroth),call(third),call(zeroth),call(second),call(zeroth),call(second),call(zeroth),call(first),call(fourth)]
+G9_calls = [call(zeroth),call(second),call(zeroth),call(zeroth),call(second),call(zeroth),call(first),call(third),call(zeroth),call(second),call(zeroth),call(second),call(zeroth),call(first),call(fourth)]
+
+leaves_to_roots_good_graphs_calls = [G0_calls,G1_calls,G2_calls,G3_calls,G4_calls,G5_calls,G6_calls,G7_calls,G8_calls,G9_calls]
+
+# Roots to Leaves recursive calls of each graph. Examples: Diagnose(fourth) recursive check_data calls.
+G0_calls = [call(fourth), call(third), call(second), call(zeroth)]
+G1_calls = [call(fourth), call(third), call(second), call(zeroth), call(first), call(zeroth), call(second), call(zeroth)]
+G2_calls = [call(fourth), call(third), call(zeroth)]
+G3_calls = [call(fourth), call(third), call(zeroth), call(second), call(zeroth)]
+G4_calls = [call(fourth), call(third), call(second), call(zeroth), call(first), call(zeroth)]
+G5_calls = [call(fourth), call(third), call(second), call(zeroth), call(first), call(zeroth), call(first), call(zeroth)]
+G6_calls = [call(fourth),call(third),call(second),call(zeroth),call(zeroth),call(first),call(second),call(zeroth),call(zeroth)]
+G7_calls = [call(fourth),call(third),call(second),call(zeroth),call(first),call(zeroth),call(zeroth),call(first),call(zeroth)]
+G8_calls = [call(fourth),call(third),call(second),call(zeroth),call(zeroth),call(second),call(zeroth),call(first),call(second),call(zeroth),call(zeroth)]
+G9_calls = [call(fourth),call(third),call(second),call(zeroth),call(zeroth),call(first),call(second),call(zeroth),call(zeroth),call(second),call(zeroth),call(first),call(second),call(zeroth),call(zeroth)]
+
+roots_to_leaves_good_graphs_calls = [G0_calls,G1_calls,G2_calls,G3_calls,G4_calls,G5_calls,G6_calls,G7_calls,G8_calls,G9_calls]
+
+# fmt: on
+
+
+##########################
+### MOCKED CONTROLLERS ###
+##########################
+class RunAutomaticCalibrationMockedController(CalibrationController):
+    """``CalibrationController`` to test the workflow of ``run_automatic_calibration()``, where its mocked ``maintain()``, ``get_last_set_parameters()`` and ``get_last_fidelities()``."""
+
+    def __init__(self, node_sequence, calibration_graph, runcard):
+        super().__init__(node_sequence=node_sequence, calibration_graph=calibration_graph, runcard=runcard)
+        self.maintain = MagicMock(return_value=None)
+        self.get_last_set_parameters = MagicMock(
+            return_value={("test", "test"): (0.0, "test", datetime.fromtimestamp(1999))}
+        )
+        self.get_last_fidelities = MagicMock(return_value={"test": (0.0, "test", datetime.fromtimestamp(1999))})
+
+
+class MaintainMockedController(CalibrationController):
```

Comment:
We'll think about it, thanks!

Reply:
Since we use each class for a different method test, we think that would be overcomplicating things, this way is very clear what we needed for each method tests!

if you still disagree, pleas make us know! Thanks! 🙌

Reply:
I do disagree indeed, that is what paramtrize is used for precisely.

Reply:
Maybe we could do it with parametrize yes!

But I don't really see any advantage compared to how its done now, and it would imply more work to do 😢 

If you really think its gives an advantage, explain it to me, and we will change it 🚀